// joltite.js/dist/util/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.apiBase = 'https://api.gamejolt.com/api/game/v1_2';
/**
 * Contains all user endpoints.
 */
const users = {
    auth: (username, token) => `/users/auth/?username=${username}&user_token=${token}`,
    fetch: (userIds) => `/users/?user_id=${userIds.join()}`,
};
/**
 * Contains all session endpoints.
 */
const sessions = {
    open: (username, token) => `/sessions/open/?username=${username}&user_token=${token}`,
    ping: (username, token, status) => `/sessions/ping/?username=${username}&user_token=${token}&status=${status}`,
    close: (username, token) => `/sessions/close/?username=${username}&user_token=${token}`,
};
/**
 * Contains all score endpoints.
 */
const scores = {
    add: (score, sort) => `/scores/add/?score=${score}&sort=${sort}`,
    fetch: () => `/scores/?`,
    tables: () => `/scores/tables/?`,
    rank: (sort) => `/scores/get-rank/?sort=${sort}`,
};
/**
 * Contains all trophy endpoints.
 */
const trophies = {
    add: (username, token, trophyId) => `/trophies/add-achieved/?username=${username}&user_token=${token}&trophy_id=${trophyId}`,
    remove: (username, token, trophyId) => `/trophies/remove-achieved/?username=${username}&user_token=${token}&trophy_id=${trophyId}`,
    fetch: (username, token) => `/trophies/?username=${username}&user_token=${token}`,
};
/**
 * Contains all data storage endpoints.
 */
const dataStorage = {
    fetch: (key) => `/data-store/?key=${key}`,
    set: (key, data) => `/data-store/set/?key=${key}&data=${data}`,
    update: (key, operation, value) => `/data-store/update/?key=${key}&operation=${operation}&value=${value}`,
    remove: (key) => `/data-store/remove/?key=${key}`,
    getKeys: () => `/data-store/get-keys/?`,
};
/**
 * Contains all friends endpoints.
 */
const friends = {
    fetch: (username, token) => `/friends/?username=${username}&user_token=${token}`,
};
/**
 * Contains all time endpoints.
 */
const time = {
    fetch: () => `/time/?`,
};
/**
 * Contains all api endpoints.
 */
exports.Endpoints = {
    users,
    sessions,
    scores,
    trophies,
    dataStorage,
    friends,
    time,
};
exports.isBrowser = typeof window !== 'undefined';
